feat [CX-2566]: Add new Generating Market Data loader for My Collection Insights

MrSltun opened the pull request:
<!--

➡️ Use a PR title in the form of  `type(PROJECT-XXXX): what changed`
➡️ Provide the Jira ticket in square brackets like [PROJECT-XXXX]

❗️ If this is a work in progress, remember to prefix it with [WIP] and/or open a draft PR instead of normal PR

-->

This PR resolves [CX-2566]

### Description

This PR adds a new loading state for fetching my collection insights after adding new artworks in My Collection

### Videos

https://user-images.githubusercontent.com/20655703/169875235-5ac15706-9510-46f5-8f0b-2eb07e605242.mp4


https://user-images.githubusercontent.com/20655703/169875250-ef280587-18b6-4332-9a61-092d8d24d722.mp4



<!-- Info, implementation, how to get there, before & after screenshots & videos, follow-up work, etc -->

### PR Checklist

- [x] I tested my changes on **iOS** / **Android**.
- [x] I added screenshots or videos to illustrate my changes.
- [x] I added Tests and Stories for my changes.
- [ ] I added an [app state migration].
- [x] I hid my changes behind a [feature flag].

### To the reviewers 👀

- [x] I would like **at least one** of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->

#### Cross-platform user-facing changes

- adding Generating Market Data loading state for My Collection - mrsltun

#### iOS user-facing changes

-

#### Android user-facing changes

-


#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[CX-2566]: https://artsyproduct.atlassian.net/browse/CX-2566?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

MrSltun commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -301,7 +332,6 @@ export const updateArtwork = async (
   }
 
   refreshMyCollection()
-  props.onSuccess?.()
```
moved this one to `handleSubmit` to stop changing the `artworkSaved` state after unmounting the artwork from

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/InsightsLoadingModal.tsx:
```diff
@@ -0,0 +1,38 @@
+  artworkSaved: boolean
+}
+
+export const InsightsLoadingModal: React.FC<InsightsLoadingModalProps & ModalProps> = ({
```
**suggestion:** I would suggest renaming this to be something more context related `SavingArtworkModal`
```suggestion
export const SavingArtworkModal: React.FC<InsightsLoadingModalProps & ModalProps> = ({
```

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/InsightsLoadingModal.tsx:
```diff
@@ -0,0 +1,38 @@
+
+  return (
+    <Modal animationType="slide" visible={isVisible} {...rest} statusBarTranslucent>
+      <View style={{ position: "absolute", width: "100%" }}>
+        <Image
+          style={{ height: screenHeight, width: "100%" }}
+          source={require("images/InsightsLoadingImage.webp")}
+          resizeMode="cover"
+        />
+      </View>
```
**suggestion:** I would instead use [ImageBackground](https://reactnative.dev/docs/imagebackground). If you need it to fill the entire screen, I do also recommend using the [predefined filling objects](https://reactnative.dev/docs/stylesheet#absolutefill-vs-absolutefillobject) from `StyleSheet`

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/InsightsLoadingModal.tsx:
```diff
@@ -0,0 +1,38 @@
+  artworkSaved = false,
+  ...rest
+}) => {
+  const { height: screenHeight } = Dimensions.get("screen")
```
**suggestion:** although we are not going to be supporting the change of the direction of the screen anytime soon, I would recommend using `height` from `useScreenDimensions()` instead to leverage our app dimensions wrapper

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/InsightsLoadingModal.tsx:
```diff
@@ -0,0 +1,38 @@
+        <Spacer my={5} />
+
+        <Text variant="md" color="white100">
+          {artworkSaved ? "Generating market data" : "Saving Artwork"}
```
**thought:** I feel a more robust API would have been providing the text as a prop instead of `artworkSaved` as a `boolean`. I don't feel like the modal needs to know that

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/InsightsLoadingModal.tsx:
```diff
@@ -0,0 +1,38 @@
+  const { height: screenHeight } = Dimensions.get("screen")
+
+  return (
+    <Modal animationType="slide" visible={isVisible} {...rest} statusBarTranslucent>
```
```suggestion
    <Modal animationType="fade" visible={isVisible} {...rest} statusBarTranslucent>
```
**issue(non-blocker):** for other loading modal, we have been using `fade` as an animation, let me know if that changed recently

olerichter00 commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -230,7 +257,11 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
           <Stack.Screen name="AddPhotos" component={MyCollectionAddPhotos} />
         </Stack.Navigator>
-        <LoadingModal isVisible={loading} />
+        {showMyCollectionInsights && props.mode === "add" ? (
+          <InsightsLoadingModal isVisible={loading || keepVisible} artworkSaved={artworkSaved} />
+        ) : (
+          <LoadingModal isVisible={loading} />
+        )}
```
I wonder if we should show the same spinner after editing as well 🤔

MrSltun replied:
I asked about it on the [slack channel](https://artsy.slack.com/archives/C01B2P6LJUU/p1652900318768949), for phase 1 we will be using the new loader only for adding new artworks

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
   formValuesRef.current = formValues
 
   const [loading, setLoading] = useState<boolean>(false)
+  const [artworkSaved, setArtworkSaved] = useState<boolean>(false)
```
**nitpick:**
```suggestion
  const [isArtworkSaved, setIsArtworkSaved] = useState<boolean>(false)
```

olerichter00 commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
-        lengthUnitPreference: preferredMetric.toUpperCase() as LengthUnitPreference,
-      })
-      await updateArtwork(values, dirtyFormCheckValues, props)
+      await Promise.all([
```
👍

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
 
   const [loading, setLoading] = useState<boolean>(false)
+  const [artworkSaved, setArtworkSaved] = useState<boolean>(false)
+  const [keepVisible, setVisible] = useState<boolean>(false)
```
**issue(blocker):** It took me a while to understand what this is referring to, how about we rename this to something like `iSavingModalVisible`

olerichter00 commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
+      if (showMyCollectionInsights) {
+        setArtworkSaved(true)
+        // simulate requesting market data
+        await new Promise((resolve) => setTimeout(resolve, 1000))
```
This will add another second to the time the modal is visible. Could we change this logic to reduce the loading time?

MrSltun replied:
this is a simulation to get the insights data, and we should change it once we have an endpoint to get the new data, or maybe remove it later in case we agree on a different way to fetch the insights data

olerichter00 reviewed (commented):
Great work 🌟

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
+        // This is to satisfy showing the insights modal for 2500 ms
+        new Promise((resolve) =>
+          setTimeout(() => {
+            setVisible(false)
```
**issue:** I think this will result in the modal being dismissed anyway after 2.5 seconds. I do though like the idea of having a promise that runs in parallel, didn't think of that before.

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
+          lengthUnitPreference: preferredMetric.toUpperCase() as LengthUnitPreference,
+        }),
+        updateArtwork(values, dirtyFormCheckValues, props),
+        // TODO: update to request for market data after saving/updating artworks
```
**thought:** I am not a big fan of having todos in the code unless they're actionable and with a ticket associated to them, I would recommend removing this.

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -230,7 +257,11 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
         </Stack.Navigator>
-        <LoadingModal isVisible={loading} />
+        {showMyCollectionInsights && props.mode === "add" ? (
+          <InsightsLoadingModal isVisible={loading || keepVisible} artworkSaved={artworkSaved} />
```
**issue:** I think `isVisible` should be only related to `isSavingModalVisible`

MounirDhahri reviewed (changes requested):
Thanks for working on this huge PR. I left some comments, please let me know if something wasn't clear. 

One thing I forgot to mention, it would be great if you also add some tests here to make sure we don't break something in the future by introducing changes here

olerichter00 commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/MyCollectionArtworkForm.tsx:
```diff
@@ -90,17 +91,42 @@ export const MyCollectionArtworkForm: React.FC<MyCollectionArtworkFormProps> = (
+      await Promise.all([
+        // This is to satisfy showing the insights modal for 2500 ms
+        new Promise((resolve) =>
+          setTimeout(() => {
+            setVisible(false)
+            return resolve("")
+          }, 2500)
```
```suggestion
          __TEST__ ? setVisible(false) : setTimeout(() => {
            setVisible(false)
            return resolve("")
          }, 2500)
```

MrSltun commented:
Thank you all for the feedback!! I updated the code and the videos in the description, and I also added tests to the code, please take a look at it now :)

dariakoko reviewed (approved):
Great work 🙌 🙌

olerichter00 reviewed (approved):
🟢

dariakoko commented:
@MrSltun , let's merge after resolving conflicts?
I think it conflicts with my changes, feel free to dm to resolve it together

MounirDhahri reviewed (approved):
🙌
